// @angular/compiler-cli/ngcc/src/execution/single_process_executor.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/ngcc/src/execution/single_process_executor", ["require", "exports", "tslib", "@angular/compiler-cli/ngcc/src/execution/utils"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var utils_1 = require("@angular/compiler-cli/ngcc/src/execution/utils");
    /**
     * An `Executor` that processes all tasks serially and completes synchronously.
     */
    var SingleProcessExecutor = /** @class */ (function () {
        function SingleProcessExecutor(logger, pkgJsonUpdater, lockFile) {
            this.logger = logger;
            this.pkgJsonUpdater = pkgJsonUpdater;
            this.lockFile = lockFile;
        }
        SingleProcessExecutor.prototype.execute = function (analyzeEntryPoints, createCompileFn) {
            var _this = this;
            this.lockFile.lock(function () {
                _this.logger.debug("Running ngcc on " + _this.constructor.name + ".");
                var taskQueue = analyzeEntryPoints();
                var compile = createCompileFn(function (task, outcome) { return utils_1.onTaskCompleted(_this.pkgJsonUpdater, task, outcome); });
                // Process all tasks.
                _this.logger.debug('Processing tasks...');
                var startTime = Date.now();
                while (!taskQueue.allTasksCompleted) {
                    var task = taskQueue.getNextTask();
                    compile(task);
                    taskQueue.markTaskCompleted(task);
                }
                var duration = Math.round((Date.now() - startTime) / 1000);
                _this.logger.debug("Processed tasks in " + duration + "s.");
            });
        };
        return SingleProcessExecutor;
    }());
    exports.SingleProcessExecutor = SingleProcessExecutor;
    /**
     * An `Executor` that processes all tasks serially, but still completes asynchronously.
     */
    var AsyncSingleProcessExecutor = /** @class */ (function (_super) {
        tslib_1.__extends(AsyncSingleProcessExecutor, _super);
        function AsyncSingleProcessExecutor() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        AsyncSingleProcessExecutor.prototype.execute = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                return tslib_1.__generator(this, function (_a) {
                    return [2 /*return*/, _super.prototype.execute.apply(this, tslib_1.__spread(args))];
                });
            });
        };
        return AsyncSingleProcessExecutor;
    }(SingleProcessExecutor));
    exports.AsyncSingleProcessExecutor = AsyncSingleProcessExecutor;
});
