// @angular/compiler-cli/ngcc/src/sourcemaps/source_file_loader.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/ngcc/src/sourcemaps/source_file_loader", ["require", "exports", "convert-source-map", "@angular/compiler-cli/src/ngtsc/file_system", "@angular/compiler-cli/ngcc/src/sourcemaps/source_file"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /**
     * @license
     * Copyright Google Inc. All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    var convert_source_map_1 = require("convert-source-map");
    var file_system_1 = require("@angular/compiler-cli/src/ngtsc/file_system");
    var source_file_1 = require("@angular/compiler-cli/ngcc/src/sourcemaps/source_file");
    /**
     * This class can be used to load a source file, its associated source map and any upstream sources.
     *
     * Since a source file might reference (or include) a source map, this class can load those too.
     * Since a source map might reference other source files, these are also loaded as needed.
     *
     * This is done recursively. The result is a "tree" of `SourceFile` objects, each containing
     * mappings to other `SourceFile` objects as necessary.
     */
    var SourceFileLoader = /** @class */ (function () {
        function SourceFileLoader(fs) {
            this.fs = fs;
        }
        SourceFileLoader.prototype.loadSourceFile = function (sourcePath, contents, mapAndPath, previousPaths) {
            if (contents === void 0) { contents = null; }
            if (mapAndPath === void 0) { mapAndPath = null; }
            if (previousPaths === void 0) { previousPaths = []; }
            if (contents === null) {
                if (!this.fs.exists(sourcePath)) {
                    return null;
                }
                // Track source file paths if we have loaded them from disk so that we don't get into an
                // infinite recursion
                if (previousPaths.includes(sourcePath)) {
                    throw new Error("Circular source file mapping dependency: " + previousPaths.join(' -> ') + " -> " + sourcePath);
                }
                previousPaths = previousPaths.concat([sourcePath]);
                contents = this.fs.readFile(sourcePath);
            }
            // If not provided try to load the source map based on the source itself
            if (mapAndPath === null) {
                mapAndPath = this.loadSourceMap(sourcePath, contents);
            }
            var map = null;
            var inline = true;
            var sources = [];
            if (mapAndPath !== null) {
                var basePath = mapAndPath.mapPath || sourcePath;
                sources = this.processSources(basePath, mapAndPath.map, previousPaths);
                map = mapAndPath.map;
                inline = mapAndPath.mapPath === null;
            }
            return new source_file_1.SourceFile(sourcePath, contents, map, inline, sources);
        };
        /**
         * Find the source map associated with the source file whose `sourcePath` and `contents` are
         * provided.
         *
         * Source maps can be inline, as part of a base64 encoded comment, or external as a separate file
         * whose path is indicated in a comment or implied from the name of the source file itself.
         */
        SourceFileLoader.prototype.loadSourceMap = function (sourcePath, contents) {
            var inline = convert_source_map_1.commentRegex.exec(contents);
            if (inline !== null) {
                return { map: convert_source_map_1.fromComment(inline.pop()).sourcemap, mapPath: null };
            }
            var external = convert_source_map_1.mapFileCommentRegex.exec(contents);
            if (external) {
                try {
                    var fileName = external[1] || external[2];
                    var externalMapPath = this.fs.resolve(this.fs.dirname(sourcePath), fileName);
                    return { map: this.loadRawSourceMap(externalMapPath), mapPath: externalMapPath };
                }
                catch (_a) {
                    return null;
                }
            }
            var impliedMapPath = file_system_1.absoluteFrom(sourcePath + '.map');
            if (this.fs.exists(impliedMapPath)) {
                return { map: this.loadRawSourceMap(impliedMapPath), mapPath: impliedMapPath };
            }
            return null;
        };
        /**
         * Iterate over each of the "sources" for this source file's source map, recursively loading each
         * source file and its associated source map.
         */
        SourceFileLoader.prototype.processSources = function (basePath, map, previousPaths) {
            var _this = this;
            var sourceRoot = this.fs.resolve(this.fs.dirname(basePath), map.sourceRoot || '');
            return map.sources.map(function (source, index) {
                var path = _this.fs.resolve(sourceRoot, source);
                var content = map.sourcesContent && map.sourcesContent[index] || null;
                return _this.loadSourceFile(path, content, null, previousPaths);
            });
        };
        /**
         * Load the source map from the file at `mapPath`, parsing its JSON contents into a `RawSourceMap`
         * object.
         */
        SourceFileLoader.prototype.loadRawSourceMap = function (mapPath) {
            return JSON.parse(this.fs.readFile(mapPath));
        };
        return SourceFileLoader;
    }());
    exports.SourceFileLoader = SourceFileLoader;
});
