// @angular/compiler-cli/ngcc/src/execution/cluster/package_json_updater.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/ngcc/src/execution/cluster/package_json_updater", ["require", "exports", "tslib", "cluster", "@angular/compiler-cli/ngcc/src/writing/package_json_updater", "@angular/compiler-cli/ngcc/src/execution/cluster/utils"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    /// <reference types="node" />
    var cluster = require("cluster");
    var package_json_updater_1 = require("@angular/compiler-cli/ngcc/src/writing/package_json_updater");
    var utils_1 = require("@angular/compiler-cli/ngcc/src/execution/cluster/utils");
    /**
     * A `PackageJsonUpdater` that can safely handle update operations on multiple processes.
     */
    var ClusterPackageJsonUpdater = /** @class */ (function () {
        function ClusterPackageJsonUpdater(delegate) {
            this.delegate = delegate;
        }
        ClusterPackageJsonUpdater.prototype.createUpdate = function () {
            var _this = this;
            return new package_json_updater_1.PackageJsonUpdate(function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                return _this.writeChanges.apply(_this, tslib_1.__spread(args));
            });
        };
        ClusterPackageJsonUpdater.prototype.writeChanges = function (changes, packageJsonPath, preExistingParsedJson) {
            var e_1, _a;
            if (cluster.isMaster) {
                // This is the master process:
                // Actually apply the changes to the file on disk.
                return this.delegate.writeChanges(changes, packageJsonPath, preExistingParsedJson);
            }
            // This is a worker process:
            // Apply the changes in-memory (if necessary) and send a message to the master process.
            if (preExistingParsedJson) {
                try {
                    for (var changes_1 = tslib_1.__values(changes), changes_1_1 = changes_1.next(); !changes_1_1.done; changes_1_1 = changes_1.next()) {
                        var _b = tslib_1.__read(changes_1_1.value, 2), propPath = _b[0], value = _b[1];
                        if (propPath.length === 0) {
                            throw new Error("Missing property path for writing value to '" + packageJsonPath + "'.");
                        }
                        // No need to take property positioning into account for in-memory representations.
                        package_json_updater_1.applyChange(preExistingParsedJson, propPath, value, 'unimportant');
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (changes_1_1 && !changes_1_1.done && (_a = changes_1.return)) _a.call(changes_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            utils_1.sendMessageToMaster({
                type: 'update-package-json',
                packageJsonPath: packageJsonPath,
                changes: changes,
            });
        };
        return ClusterPackageJsonUpdater;
    }());
    exports.ClusterPackageJsonUpdater = ClusterPackageJsonUpdater;
});
